cmd/clusterctl/pkg/log/logger.go

```diff
@@ -0,0 +1,278 @@
+
+Logging errors:
+
+A proper error management should always be preferred to the usage of log.Error.
+
+*/
+
+package log
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+)
+
+// LogEntry defines the information that can be used for composing a log line.
+type LogEntry struct {
+	// Prefix of the log lines, composed by the hierarchy of log.WithName values.
+	Prefix string
+
+	// Level of the LogEntry.
+	Level int
+
+	// Values of the log lines, composed by the concatenation of log.WithValues and KeyValue pairs passed to lg.Info/log..
+	Values []interface{}
+}
+
+// LogChannel define a channel where LogEntry can be sent.
+type LogChannel chan LogEntry
+
+// NewLogger returns a new instance of the clusterctl logger.
+func NewLogger() logr.Logger {
+	return &Logger{
+		level:  0,
+		prefix: "",
+		values: nil,
+	}
+}
+
+// NewLogger returns a new instance of the clusterctl logger ready for sending log entries to a channel.
+func NewLoggerWithChannel(ch LogChannel) logr.Logger {
+	return &Logger{
+		ch:     ch,
+		level:  0,
+		prefix: "",
+		values: nil,
+	}
+}
+
+// Logger defines a clusterctl friendly logr.Logger
+type Logger struct {
+	ch     LogChannel
+	level  int
+	prefix string
+	values []interface{}
+}
+
+var _ logr.Logger = &Logger{}
+
+// Enabled tests whether this Logger is enabled.
+// NB. This method relies on klog settings.
+func (l *Logger) Enabled() bool {
+	return bool(klog.V(klog.Level(l.level)))
+}
+
+// Info logs a non-error message with the given key/value pairs as context.
+// NB. This method relies on klog settings.
+func (l *Logger) Info(msg string, kvs ...interface{}) {
+	if l.Enabled() {
+		values := copySlice(l.values)
+		values = append(values, kvs...)
+		values = append(values, "msg", msg)
+		l.write(values)
+	}
+}
+
+// Error logs an error message with the given key/value pairs as context.
+func (l *Logger) Error(err error, msg string, kvs ...interface{}) {
+	values := copySlice(l.values)
+	values = append(values, kvs...)
+	values = append(values, "msg", msg, "error", err)
+	l.write(values)
+}
+
+// V returns an InfoLogger value for a specific verbosity level.
+func (l *Logger) V(level int) logr.InfoLogger {
+	new := l.clone()
+	new.level = level
+	return new
+}
+
+// WithName adds a new element to the logger's name.
+func (l *Logger) WithName(name string) logr.Logger {
+	new := l.clone()
+	if len(l.prefix) > 0 {
+		new.prefix = l.prefix + "/"
+	}
+	new.prefix += name
+	return new
+}
+
+// WithValues adds some key-value pairs of context to a logger.
+func (l *Logger) WithValues(kvList ...interface{}) logr.Logger {
+	new := l.clone()
+	new.values = append(new.values, kvList...)
+	return new
+}
+
+func (l *Logger) write(values []interface{}) {
+	entry := LogEntry{
+		Prefix: l.prefix,
+		Level:  l.level,
+		Values: values,
+	}
+	if l.ch != nil {
+		l.ch <- entry
```

Comment:
This will block if the channel doesn't have an active reader and/or if the channel doesn't have sufficient capacity. Do you want to drop log messages in that case, or block?

Reply:
I'm fine with blocking for now

Reply:
I don't know if blocking is the right behavior. After working in observability for a while, a rule we followed was not having the logging framework impact the usability of the application. I suggest a non-blocking send at the very least here. It feels like the user may misunderstand the usage of `NewLoggerWithChannel`.